libfwupdplugin/fu-block-device.c

```diff
@@ -0,0 +1,395 @@
+/*
+ * Copyright (C) 2021 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN "FuBlockDevice"
+
+#include "config.h"
+
+#include "fu-block-device.h"
+#include "fu-common.h"
+#include "fu-io-channel.h"
+
+/**
+ * FuBlockDevice:
+ *
+ * A block device, typically a FAT32 volume
+ *
+ * See also: [class@FuDevice]
+ */
+
+typedef struct {
+	gchar *uuid;
+	gchar *label;
+	gchar *filename;
+} FuBlockDevicePrivate;
+
+enum { PROP_0, PROP_UUID, PROP_LABEL, PROP_FILENAME, PROP_LAST };
+
+G_DEFINE_TYPE_WITH_PRIVATE(FuBlockDevice, fu_block_device, FU_TYPE_DEVICE)
+
+#define GET_PRIVATE(o) (fu_block_device_get_instance_private(o))
+
+#define FU_BLOCK_DEVICE_SIZE_MAX     16 * 1024 * 1024
+#define FU_BLOCK_DEVICE_READ_TIMEOUT 15 * 1000
+
+static void
+fu_block_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuBlockDevice *self = FU_BLOCK_DEVICE(device);
+	FuBlockDevicePrivate *priv = GET_PRIVATE(self);
+	fu_common_string_append_kv(str, idt, "Uuid", priv->uuid);
+	fu_common_string_append_kv(str, idt, "Label", priv->label);
+	fu_common_string_append_kv(str, idt, "Filename", priv->filename);
+}
+
+static gboolean
+fu_block_device_probe(FuDevice *device, GError **error)
+{
+	FuBlockDevice *self = FU_BLOCK_DEVICE(device);
+	FuBlockDevicePrivate *priv = GET_PRIVATE(self);
+
+	if (priv->uuid != NULL) {
+		g_autofree gchar *instance_id0 = NULL;
+		instance_id0 = g_strdup_printf("BLOCK\\UUID_%s", priv->uuid);
+		fu_device_add_instance_id(FU_DEVICE(self), instance_id0);
+	}
+	if (priv->label != NULL) {
+		g_autofree gchar *instance_id0 = NULL;
+		instance_id0 = g_strdup_printf("BLOCK\\LABEL_%s", priv->label);
```

Comment:
I'm a bit worried about instances that are entirely built on a label or UUID.  need more factors in these UUIDs and/or drop them.

* UUID: Too specific, if it's reformatted you can no longer target it.
* Label: Too generic.  It's just asking for me to be installing updates for my camera to my OS recovery partition if they're both called `RECOVERY`

Reply:
> UUID: Too specific, if it's reformatted you can no longer target it.

I actually think UUID is okay; it's not possible to "format" a UF2 drive, it's just ignored. It's not really a filesystem, it's just a driver that reads the 512 byte blocks and writes them to each sector. If you formatted it as ext4 for instance it'd come back as a FAT32 drive with the hardcoded UUID.

Label is tricky, I agree. How about `BLOCK\\UUID_%s` and `BLOCK\\UUID_%s&LABEL_%s"?

Reply:
`INFO_UF2.TXT` should have a bunch of information - that is what should be going into the instance IDs, not this stuff in my mind.

Reply:
> that is what should be going into the instance IDs

Right, but that's one layer "up" -- this is the `FuBlockDevice` that lives in libfwupdplugin (that we match on in the quirk file) and then the `FuUf2Device` that superclasses the `FuBlockDevice` that adds the `INFO_UF2.TXT`-derived information for matching the firmware stream.

Reply:
> Right, but that's one layer "up
Oh right.

>and then the FuUf2Device that superclasses the FuBlockDevice that adds the INFO_UF2.TXT-derived information for matching the firmware stream

Even being the wrong layer - I don't think we want these UUIDs to be anything that can be used to target a device for the same generic or too specific reasons I mentioned.  I think that step should just be skipped.

Reply:
> I think that step should just be skipped.

Well, we need to match the device by UUID or LABEL so we know if we should create the FuBlockDevice and offer it to plugins -- I really don't think we want to be reading from *all* random VFAT drives as soon as they're plugged in -- that seems like the kind of change that would make people upset. I think fwupd looking for a `INFO_UF2.TXT` on a drive called `FWUPDATE` or where the UUID is allow-listed by a vendor is probably acceptable. I suppose one compromise would be to add the `BLOCK\\UUID_%s` instance IDs as `FU_DEVICE_INSTANCE_FLAG_ONLY_QUIRKS` so you can't use them to match to a firmware stream.